vms/platformvm/state/merkle_state.go

```diff
@@ -0,0 +1,1530 @@
+	lastAcceptedHeight uint64                        // TODO: Should this be written to state??
+	supplies           map[ids.ID]uint64             // map of subnetID -> current supply
+	suppliesCache      cache.Cacher[ids.ID, *uint64] // cache of subnetID -> current supply if the entry is nil, it is not in the database
+
+	// Subnets section
+	addedPermissionedSubnets []*txs.Tx                     // added SubnetTxs, waiting to be committed
+	permissionedSubnetCache  []*txs.Tx                     // nil if the subnets haven't been loaded
+	addedElasticSubnets      map[ids.ID]*txs.Tx            // map of subnetID -> transformSubnetTx
+	elasticSubnetCache       cache.Cacher[ids.ID, *txs.Tx] // cache of subnetID -> transformSubnetTx if the entry is nil, it is not in the database
+
+	// Chains section
+	addedChains map[ids.ID][]*txs.Tx            // maps subnetID -> the newly added chains to the subnet
+	chainCache  cache.Cacher[ids.ID, []*txs.Tx] // cache of subnetID -> the chains after all local modifications []*txs.Tx
+
+	// Blocks section
+	addedBlocks map[ids.ID]blocks.Block            // map of blockID -> Block
+	blockCache  cache.Cacher[ids.ID, blocks.Block] // cache of blockID -> Block. If the entry is nil, it is not in the database
+	blockDB     database.Database
+
+	// Txs section
+	// FIND a way to reduce use of these. No use in verification of addedTxs
+	// a limited windows to support APIs
+	addedTxs map[ids.ID]*txAndStatus            // map of txID -> {*txs.Tx, Status}
+	txCache  cache.Cacher[ids.ID, *txAndStatus] // txID -> {*txs.Tx, Status}. If the entry is nil, it isn't in the database
+	txDB     database.Database
+
+	indexedUTXOsDB database.Database
+
+	localUptimesCache    map[ids.NodeID]map[ids.ID]*uptimes // vdrID -> subnetID -> metadata
+	modifiedLocalUptimes map[ids.NodeID]set.Set[ids.ID]     // vdrID -> subnetIDs
+	localUptimesDB       database.Database
+
+	validatorWeightDiffsCache cache.Cacher[heightWithSubnet, map[ids.NodeID]*ValidatorWeightDiff] // heightWithSubnet -> map[ids.NodeID]*ValidatorWeightDiff
+	localWeightDiffDB         database.Database
+
+	validatorBlsKeyDiffsCache cache.Cacher[uint64, map[ids.NodeID]*bls.PublicKey] // cache of height -> map[ids.NodeID]*bls.PublicKey
+	localBlsKeyDiffDB         database.Database
+}
+
+// STAKERS section
+func (ms *merkleState) GetCurrentValidator(subnetID ids.ID, nodeID ids.NodeID) (*Staker, error) {
+	return ms.currentStakers.GetValidator(subnetID, nodeID)
+}
+
+func (ms *merkleState) PutCurrentValidator(staker *Staker) {
+	ms.currentStakers.PutValidator(staker)
+
+	// make sure that each new validator has an uptime entry
+	// and a delegatee reward entry. MerkleState implementations
+	// of SetUptime and SetDelegateeReward must not err
+	err := ms.SetUptime(staker.NodeID, staker.SubnetID, 0 /*duration*/, staker.StartTime)
+	if err != nil {
+		panic(err)
+	}
+	err = ms.SetDelegateeReward(staker.SubnetID, staker.NodeID, 0)
+	if err != nil {
+		panic(err)
+	}
```

Comment:
not neat at all, but an invariant I rediscovered, which is not super clear in state.State (where it's handled via metadata struct)